### ISSUE TITLE
Leaving DS4Windows connected and returning requires a PC restart before working

### ISSUE THREAD

**Issue by rincebrain** [Ryochan7/DS4Windows#506]
I've had this at least with 1.5.8, 1.5.15, 1.5.16, and 1.5.17.

On clean boot of my Win10 machine, I start DS4Windows, hold the PS button on my controller to turn it on, and everything works wonderfully.

If I leave DS4Windows running for an extended interval with my PC locked, and the controller connected, the controller will appear to work, but games will not actually see any inputs from it.

The games are configured to use the X360 Controller.

The Game Controller Properties -> Test input for the X360 Controller sees all the correct button inputs, DS4Windows's own interface for keybinds sees everything correctly, but the game I'm usually trying to play when this happens (Final Fantasy 14, not purchased or launched through Steam) sees none of them.

Using the Stop/Start button in DS4Windows doesn't help, closing FF14 or DS4Windows (and making sure there's not a process hanging around in Task Manager), disconnecting+reconnecting the controller, or trying to connect it via microUSB (I usually use it wirelessly), none of it helps.

Rebooting resolves it every time.

Attached is the Log tab's output when I left it running for the past 26 hours or so. It was functioning perfectly fine on a fresh reboot, but is not, any more, and is not logging any obvious bloody fingers.
[ds4 silly behavior log.txt](https://github.com/Ryochan7/DS4Windows/files/2641776/ds4.silly.behavior.log.txt)

**Comment by Ryochan7** [comment]
What is going on in the background that is causing the latency warning to keep going off? There does not seem to be anything else odd in the log file.

I have run DS4Windows overnight on many occasions and I have not run into this kind of problem. Locking the computer should not make a difference but I will try it out later just in case. If inputs are still detected in the Game Controller Properties window then ScpVBus is still doing its job so the problem is not with the output driver. I wouldn't know why FF14 would have a problem reading controller inputs when other applications can. Exclusive mode is still active so it is not like the game is attempting to use the original DirectInput device instead of the virtual XInput device. If exclusive mode failed at any point then the problem would make sense.

**Comment by rincebrain** [comment]
I don't actually know what's causing the latency spikes occasionally; it's not been visible from a gameplay perspective so I've just been ignoring it. My guess would either be some tiny stutter on clockspeed shift or the USB Bluetooth dongle choking, since I don't have this particular machine set to not try power saving. Could probably instrument it with ETW and find out, but as it's not been user-visible, it's also not been a priority.

Yeah, my first guess was something was disconnected between the real and virtual devices, but since Windows sees all the X360 buttons being pressed, it's not clear to me where something _could_ go awry, unless it's doing something strange with keybindings. I had thought I'd found the issue when I found an older version of DS4Windows was being launched when I reconnected a controller, but having removed that, the issue does persist.

I've got two profiles set up, one with DInput only and one without (the default), so I could swap to DInput only and back again when I had inconsistent misbehaviors in much older versions, but that's never been helpful with this particular issue. No special-casing of any programs is setup.

**Comment by mika-n** [comment]
Have you setup a linking between a profile and controller? Or auto-switching of a profile based on a running application? Sometimes I have seen that auto-switching doesn't pickup the right profile when computer comes back alive from a sleep. You said you have two profiles, one with DInput only and other profile using the normal XInput for x360. Maybe your setup picks up a wrong profile if you have auto-switching on?  Try without auto-switching.

**Comment by rincebrain** [comment]
I have no auto-switching for specific programs, I do have a default set for a particular controller, but since I have them display two different controller LED colors when they load, unless you think it's half-setting details from multiple profiles, I'm reasonably confident that isn't the case?

Plus, switching profiles then switching back doesn't fix it.

**Comment by rincebrain** [comment]
It's doing it again; I have Visual Studio hanging around, so if you'd like any specific information or poking to figure out what's going on, I can probably manage it.

**Comment by rincebrain** [comment]
I had a better idea, and just took full process dumps from when it was not working after being left running for a while, when it was not working after being entirely closed and launched anew, and when it was working after rebooting from that.

Admittedly, it's not also a kernel dump, mini or full, but since I don't have the environment set up to interactively take one of those over firewire, ethernet, or serial at the moment, it'll have to do.

The unnumbered one is when it was running for a while and didn't work.
The one numbered (2) is when it still didn't work after being closed and launched fresh without rebooting.
The one numbered (3) is when it worked perfectly when launched anew after rebooting.

https://www.dropbox.com/sh/0036ppt1xmp6az1/AAAHk_enU98XW__5sXQmG7BNa?dl=0

**Comment by Ryochan7** [comment]
From what I can decipher from the dumps, DS4Windows seems to be operating fine. Each dump shows that the input thread is waiting for the next input packet from the DS4 device. The other threads related to the device seem to be at an expected state as well. Since other tools show that input packets are still flowing, even when problems occur, then the thread is working as expected.

```
HidDevice.ReadStatus res = hDevice.ReadWithFileStream(btInputReport);
```

I am stumped at this point. There is a 30 day trial demo of Final Fantasy 14 so I will try it out. Unless some other tool is causing a problem or there is some kind of anti-hack protection in place that might cause a problem then I don't think I will encounter the problem on my end.

**Comment by Ryochan7** [comment]
Left my computer running in locked mode for 8 hours with DS4Windows running and a controller connected the whole time. Went back to FF 14 and it worked.

One thing that I just thought of that should be mentioned is that DInput only mode is pretty much useless for a BT connection. Due to the mode change that DS4Windows performs on the DS4, the original DirectInput Wireless Controller device will be seen as a dead controller to any application that tries to use it.  DInput only mode works for USB and Sony Wireless Adapter connections for allowing games to use the DS4 natively rather than as an emulated Xbox 360 controller. DInput only mode for BT connections only allows using exclusive KB+M bindings for games that have problems with mixed input controls.

**Comment by rincebrain** [comment]
Yeah, I can't always reproduce it, whether I leave the controller plugged in via microUSB to charge, unplugged but connected, or unplugged and unconnected. But when it does reproduce, it won't go away prior to reboot.

I've spent a bit trying to figure out where the disconnect could possibly be, but none of the answers work unless something's getting loaded into FF14 and somehow mangling the controller input support with 100% fidelity once initially triggered. :/

**Comment by Ryochan7** [comment]
I can't reproduce the problem on my end. It is a cuck move but here is an updated ViGEm build. You can try using the ViGEm build if you have ViGEmBus installed. Maybe you would have better luck with FF 14 and ViGEm than using ScpVBus. I have been using ScpVBus as the output driver during my testing.

http://ryochan7.com/ds4windows/test/DS4Windows_1.5.18_ViGEm_UdpServer_x64.zip

**Comment by Ryochan7** [comment]
I have only tried this once more since and I have not run into a problem. This has only shown me that I would not have the time to invest in a real MMORPG. I would be more interesting in a lite MMO like Phantasy Star Online.

**Comment by Ryochan7** [comment]
It has been one month so my demo period is over.

**Comment by rincebrain** [comment]
I was out of town (and consequently, away from controller) for a couple of weeks, and haven't reproduced it since I returned.

Sample size small because, well, it's been about a week, but.

**Comment by rincebrain** [comment]
It came back once and as before, rebooting resolved it, but nothing else did.

I think it's a problem with the ScpVbus driver getting stuck in a weird state, though, because I just came home to a new failure mode - I start FF14, it displays a black windowed screen Not Responding until I kill it, on repeated instances...but the moment I close DS4Windows it runs normally. Starting DS4Windows again after that, it seems to loop on trying to connect to the ScpVbus driver and thinking it got Exclusive Mode, then losing connection - the button in the bottom-right always says "Start" and doesn't appear to change when clicked. Windows wouldn't uninstall the Xbox 360 controller (it just blocked forever), and  hung indefinitely on "Restarting..." until I powered it off.

So I think something's rotten in the state of Denmark there.

(Not expecting you to do anything about it, just sharing in case someone finds this.)

**Comment by Ryochan7** [comment]
It is too bad that the last version of the ScpVBus driver (1.0.0.103) is getting close to 6 years old now (2013/05/05) and ViGEmBus had not been a suitable replacement in my opinion. More problems are bound to come up as new Windows updates are released and there is not much of a course of action to fix any problems that will arise. It is too bad that Microsoft puts this type of problem behind an expensive paywall which kills any real possibility of distributing any improvements.

**Comment by rincebrain** [comment]
Forgive my ignorance, but isn't the latest release on
https://github.com/nefarius/ScpVBus/releases only 3 years old? I don't
think I know the history of this project and similarly named ones.
Admittedly, it didn't build when I asked Visual Studio nicely.

- Rich

**Comment by Ryochan7** [comment]
I have only used that version of ScpVBus when I tried out ScpToolkit back when I first got my DS4 controller. I did not like the performance of ScpToolkit when I used it so I don't think I would probably want to use that version of ScpVBus. Other mapping programs that I use rely on the older Scarlet Crush version which is what DS4Windows uses as well; InputMapper 1.6 and Touchmote are the main programs that I use and bundle the Scarlet Crush build.

Either way, that version is not supported anymore either. ViGEmBus support was supposed to be used by DS4Windows but things broke down despite having a working build. With a custom built version of ViGEmBus with custom build options, ViGEmBus does outperform ScpVBus. But other people will be relying on the current public release. I tried to help with the project but that didn't go anywhere.
